File: members/U02CDR60PFS/kakao2021/순위_검색.js
function solution(info, query) {
  let answer = [];
  const infos = {};

  info = info.map((e) => {
    e = e.split(" ");
    const score = e.pop() * 1;
    dfs(0, []);

    function dfs(idx, info) {
      const key = info.join("");
      infos[key] = infos[key] || [];
      infos[key].push(score);

      for (let i = idx; i < e.length; i++) {
        info.push(e[i]);
        dfs(i + 1, info);
        info.pop();
      }
    }
  });

  Object.keys(infos).forEach((e) => infos[e].sort((a, b) => a - b));

  query.forEach((q) => {
    q = q.split(" ");
    const score = q.pop();
    q = q.filter((e) => e !== "-" && e !== "and").join("");
    if (!infos[q]) {
      answer.push(0);
      return;
    }

    let cnt = binarySearch(0, infos[q].length, score);
    answer.push(cnt);

    function binarySearch(left, right, target) {
      while (left < right) {
        let mid = parseInt((left + right) / 2);
        if (infos[q][mid] < target) {
          left = mid + 1;
        } else if (infos[q][mid] >= target) {
          right = mid;
        }
      }
      return infos[q].length - left;
    }
  });
  return answer;
}

File: members/U02BQ6G1SQJ/programmers/problem/순위 검색.py
# 소요 시간 : 49분
import itertools;

LANGUAGE_LIST = ["-", "cpp", "java", "python"];
JOB_LIST = ["-", "backend", "frontend"];
CAREER_LIST = ["-", "junior", "senior"];
FOOD_LIST = ["-", "chicken", "pizza"];

def solution(info, query):
    (answerList, infoList) = ([], []);
    productList = list(itertools.product(LANGUAGE_LIST, JOB_LIST, CAREER_LIST, FOOD_LIST));
    queryDict = dict(zip(productList, [[] for _ in range(len(productList))]));
    
    for infoStr in info:
        (language, job, career, food, score) = infoStr.split(" ");
        
        # print(language, job, career, food, score);
        
        for p in itertools.product(["-", language], ["-", job], ["-", career], ["-", food]):
            # print(p);
            
            queryDict[p].append(int(score));
    
    for queryVal in queryDict.values():
        queryVal.sort();
        
    # print(infoList);
    # print(productList);
    # print(queryDict);
    
    for queryStr in query:
        queryList = queryStr.split(" ");
        (queryKey, score) = ((queryList[0], queryList[2], queryList[4], queryList[6]), int(queryList[7]));
        (answer, left, right) = (len(queryDict[queryKey]), 0, len(queryDict[queryKey]) - 1);
        
        while (left <= right):
            mid = (left + right) // 2;
            
            # print(mid, left, right);
            # print(queryDict[queryKey][mid], queryDict[queryKey][left], queryDict[queryKey][right]);
            
            if (queryDict[queryKey][mid] < score):
                left = mid + 1;
            else:
                (answer, right) = (mid, mid - 1);
        
        answerList.append(len(queryDict[queryKey]) - answer);
    
    return answerList;
